flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractFlinkResourceReconciler.java

```diff
@@ -99,15 +103,20 @@ public final void reconcile(CR cr, Context ctx) throws Exception {
         // No further logic is required at this point.
         if (firstDeployment) {
             LOG.info("Deploying for the first time");
+
+            // Before we try to submit the job we record the current spec in the status so we can
+            // handle subsequent deployment and status update errors
+            ReconciliationUtils.updateStatusBeforeDeploymentAttempt(cr, deployConfig);
+            statusRecorder.patchAndCacheStatus(cr);
+
             deploy(
                     cr,
                     spec,
                     status,
                     deployConfig,
                     Optional.ofNullable(spec.getJob()).map(JobSpec::getInitialSavepointPath),
                     false);
-            ReconciliationUtils.updateForSpecReconciliationSuccess(
-                    cr, JobState.RUNNING, deployConfig);
+            ReconciliationUtils.updateStatusForDeployedSpec(cr, deployConfig);
```

Comment:
If we throw exception in the deploy, the resource may not be truly deployed. So, I think the `firstDeployment` condition is not full. IMO, It should includes :

- lastReconciledSpec == null
- Or state in the `ReconciliationState.UPGRADING` and `JobManagerDeploymentStatus.Missing`

Reply:
I'm second to @morhidi  to make the `isFirstDeploy` as a dedicated methods.

Reply:
@Aitozi it is only safe to directly deploy if the lastReconciledSpec was null, otherwise it might be suspended due to a spec change etc. I think the current logic is correct.

Reply:
I think the problem around the upgrade can be fixed by this PR. The only weak part is about the firstDeploy. Because the upgrade failure will be taken over by the check during the next reconciler for 

```
 boolean specChanged =
                reconciliationStatus.getState() == ReconciliationState.UPGRADING
                        || !currentDeploySpec.equals(lastReconciledSpec);
```

But the firstDeploy do not have the same thing. What about introduce the `INITIALIZING` for the firstDeploy to distinguish between the upgrade ?

Reply:
Because this case is covered I would not add a new ReconciliationState unless necessary

Reply:
Then, we can know whether it is an firstDeploy by `lastReconciledSpec == null ||  reconciliationStatus.getState() == ReconciliationState.INITIALIZING`

Reply:
@Aitozi I just manually verified this again by throwing errors directly before / after first deploy. And it both works.